lib/src/revset_graph.rs

```diff
@@ -172,6 +180,29 @@ where
         Some(())
     }
 
+    fn flush_new_heads(&mut self) {
+        assert!(!self.new_head_ids.is_empty());
+        if self.blocked_ids.is_empty() || self.new_head_ids.len() <= 1 {
+            // Orphaned (or no choice), pick the first head
+            self.emittable_ids.push(self.new_head_ids.remove(0));
+        } else {
+            // Queue heads which are reachable from the emittable sub graph
+            let mut to_visit: Vec<&CommitId> = self.blocked_ids.iter().collect();
+            let mut reached: HashSet<&CommitId> = self.blocked_ids.iter().collect();
+            while let Some(id) = to_visit.pop() {
+                let node = self.nodes.get(id).unwrap();
+                to_visit.extend(node.child_ids.iter().filter(|id| reached.insert(id)));
+            }
+            // TODO: Use drain_filter() if gets stabilized
+            let (new_reachable, new_unreachable) = mem::take(&mut self.new_head_ids)
+                .into_iter()
+                .partition::<Vec<_>, _>(|id| reached.contains(id));
+            assert!(!new_reachable.is_empty(), "blocking head should exist");
+            self.emittable_ids.extend(new_reachable.into_iter().rev());
+            self.new_head_ids = new_unreachable;
```

Comment:
Perhaps, we'll need to pick one head to queue in. Reachable heads may be arbitrary ordered.